cmd/agent/app/jmx.go

```diff
@@ -0,0 +1,184 @@
+		Long:  ``,
+		Run:   doJmxListCollected,
+	}
+
+	jmxListNotMatchingCmd = &cobra.Command{
+		Use:   "not-matching",
+		Short: "List attributes that don’t match any of your instances configuration.",
+		Long:  ``,
+		Run:   doJmxListNotCollected,
+	}
+
+	checks = []string{}
+)
+
+func init() {
+	// attach list and collect commands to jmx command
+	jmxCmd.AddCommand(jmxListCmd)
+	jmxCmd.AddCommand(jmxCollectCmd)
+
+	//attach list commands to list root
+	jmxListCmd.AddCommand(jmxListEverythingCmd, jmxListMatchingCmd, jmxListLimitedCmd, jmxListCollectedCmd, jmxListNotMatchingCmd)
+
+	jmxListCmd.PersistentFlags().StringSliceVar(&checks, "checks", []string{"jmx"}, "JMX checks (ex: jmx,tomcat)")
+	jmxCollectCmd.PersistentFlags().StringSliceVar(&checks, "checks", []string{"jmx"}, "JMX checks (ex: jmx,tomcat)")
+
+	// attach the command to the root
+	AgentCmd.AddCommand(jmxCmd)
+}
+
+func doJmxCollect(cmd *cobra.Command, args []string) {
+	runJmxCommand("collect")
+}
+
+func doJmxListEverything(cmd *cobra.Command, args []string) {
+	runJmxCommand("list_everything")
+}
+
+func doJmxListMatching(cmd *cobra.Command, args []string) {
+	runJmxCommand("list_matching_attributes")
+}
+
+func doJmxListLimited(cmd *cobra.Command, args []string) {
+	runJmxCommand("list_limited_attributes")
+}
+
+func doJmxListCollected(cmd *cobra.Command, args []string) {
+	runJmxCommand("list_collected_attributes")
+}
+
+func doJmxListNotCollected(cmd *cobra.Command, args []string) {
+	runJmxCommand("list_not_matching_attributes")
+}
+
+func runJmxCommand(command string) {
+	err := common.SetupConfig(confFilePath)
+	if err != nil {
+		log.Fatalln("unable to set up global agent configuration: %v", err)
+	}
+
+	common.SetupAutoConfig(config.Datadog.GetString("confd_path"))
+
+	runner := jmxfetch.New()
+
+	runner.ReportOnConsole = true
+	runner.Command = command
+
+	dir, err := ioutil.TempDir(os.TempDir(), "jmxconfd")
+	if err != nil {
+		log.Fatalln(err)
+	}
+	defer os.RemoveAll(dir)
+	runner.ConfDirectory = dir
+
+	configs := common.AC.GetAllConfigs()
+
+	for _, c := range checks {
+		config, err := findJMXConfigByCheckName(configs, c)
+		if err != nil {
+			log.Fatalln(err)
+		}
+
+		file, err := ioutil.TempFile(dir, fmt.Sprintf("conf_%v", c))
+		if err != nil {
+			log.Fatalln(err)
+		}
+		defer file.Close()
+
+		_, err = file.WriteString(config.String())
+		if err != nil {
+			log.Fatalln(err)
+		}
+
+		runner.Checks = append(runner.Checks, filepath.Base(file.Name()))
+	}
+
+	err = runner.Run()
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	err = runner.Wait()
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	fmt.Println("JMXFetch exited successfully. If nothing was displayed please check your configuration, flags and the JMXFetch log file.")
+}
+
+func findJMXConfigByCheckName(configs []check.Config, checkName string) (*check.Config, error) {
+	for _, c := range configs {
+		if strings.EqualFold(c.Name, checkName) {
+			if c.IsJMX() {
+				return &c, nil
```

Comment:
Unhandled edge case: unless I'm missing something, `configs` may contain multiple configs for a given check name. This would only take into account the first one.